### PR TITLE
Fix listener leave event order

### PR DESCRIPTION
**Pull request by kddnewton**

### CODE REVIEW

**Comment by vinistock** [comment]
Looks great! This fixes the ordering issue with events.

**Review by jemmaissroff — APPROVED** [review]

File: templates/lib/yarp/node.rb.erb [hunk 1]
```diff
@@ -210,15 +240,19 @@ module YARP
...
+      # contains (event, node) pairs.
+      while (event, node = queue.shift)
+        case event
         <%- nodes.each do |node| -%>
-        when :<%= node.human %>
-          listeners[:<%= node.human %>_enter]&.each { |listener| listener.<%= node.human %>_enter(node) }
-          queue = node.compact_child_nodes.concat(queue)
-          listeners[:<%= node.human %>_leave]&.each { |listener| listener.<%= node.human %>_leave(node) }
+        when :on_<%= node.human %>_enter
+          listeners[:on_<%= node.human %>_enter]&.each { |listener| listener.on_<%= node.human %>_enter(event) }
```
Comment (vinistock) [inline templates/lib/yarp/node.rb.erb:251]:
```suggestion
          listeners[:on_<%= node.human %>_enter]&.each { |listener| listener.on_<%= node.human %>_enter(node) }
```

File: templates/lib/yarp/node.rb.erb [hunk 2]
```diff
@@ -210,15 +240,19 @@ module YARP
...
     def dispatch(root)
-      queue = [root]
+      queue = [[root.enter_event, root]]
 
-      while (node = queue.shift)
-        case node.human
+      # This inline queue avoids having to recursively walk the tree. The queue
+      # contains (event, node) pairs.
+      while (event, node = queue.shift)
+        case event
```
Comment (vinistock) [inline templates/lib/yarp/node.rb.erb:248]: We also need to fire a generic `on_node_enter` and `on_node_leave` for all nodes.

File: templates/lib/yarp/node.rb.erb [hunk 3]
```diff
@@ -210,15 +240,19 @@ module YARP
     #
     # def dispatch: (Node) -> void
     def dispatch(root)
-      queue = [root]
+      queue = [[root.enter_event, root]]
 
-      while (node = queue.shift)
-        case node.human
+      # This inline queue avoids having to recursively walk the tree. The queue
+      # contains (event, node) pairs.
```
[the remark this answers is not in the document]
Comment (eregon) [reply templates/lib/yarp/node.rb.erb:246]: What do you think about this?

[the remark this answers is not in the document]
Comment (kddnewton) [reply templates/lib/yarp/node.rb.erb]: Yeah, I've updated it to be recursive.

File: templates/lib/yarp/node.rb.erb (hunk 2 above)
Comment (kddnewton) [reply templates/lib/yarp/node.rb.erb]: Let's do that in a follow-up PR, I want to get this merged

File: templates/lib/yarp/node.rb.erb (hunk 3 above)
[the remark this answers is not in the document]
Comment (eregon) [reply templates/lib/yarp/node.rb.erb]: Great, it's so much simpler now and the event order is naturally the same as the Visitor.

**Pull request merged**